Add QIT workflow

tomalec opened the pull request:
:warning: depends on https://github.com/woocommerce/grow/pull/81

### Changes proposed in this Pull Request:

1. Change the `build` workflow so it publish an artifact and can be called by other workflows.
2. Add a manual workflow, that builds the extension and run selected QIT tests for that package.


### Screenshots:

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17435/b6674b1c-32a6-4b71-b681-5b1b058b9b1e)



### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Fork the repo
2. Setup secrets. Links to secrets are at Pe98dy-Ir-p2 
3. Merge this branch to the main one /`trunk`
4. Run the manual workflow at  https://github.com/woocommerce/google-listings-and-ads/actions/workflows/run-qit.yml




### Additional details:

1. Before margin we need to remove those lines 
https://github.com/woocommerce/google-listings-and-ads/pull/2114/files#diff-0550f940c28a3fbd24937bad01914ede672b538aff10de709e999759ccf4f9dfR41-R44

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Dev - Add manual QIT workflow

tomalec committed:
Add QIT workflow

tomalec committed:
Require build to run qit workflow.

puntope commented on .github/workflows/run-qit.yml:
```diff
@@ -0,0 +1,66 @@
+        required: true
+        default: true
+        type: boolean
+      test-e2e:
+        description: 'Should E2E be tested? (takes a lot of time)'
+        required: true
+        default: false
+        type: boolean
```
Are not the e2e tests already in another workflow?

tomalec replied:
These are QIT E2E tests, not local GLA E2Es.

`qit run:e2e` run the suite of E2E tests to check if installing GLA does not break WC core behavior.

[`.github/workflows/e2e-tests.yml`](https://github.com/woocommerce/google-listings-and-ads/blob/develop/.github/workflows/e2e-tests.yml) (if that's the workflow you mean) - tests GLA-specific behavior

puntope commented on .github/workflows/run-qit.yml:
```diff
@@ -0,0 +1,66 @@
+      test-activation: ${{ inputs.test-activation || true }}
+      test-security: ${{ inputs.test-security || true }}
+      test-phpstan: ${{ inputs.test-phpstan || true }}
+      test-api: ${{ inputs.test-api || false }}
```
Test API is true by default in the inputs. Shouldn't be this consistent with it, as the rest of the parameters?

puntope commented on .github/workflows/run-qit.yml:
```diff
@@ -0,0 +1,66 @@
+      - dev/qit-workflow
+
+jobs:
+  build:
+    name: Build extension
+    uses: ./.github/workflows/build.yml
+    secrets: inherit
```
If we generated the artifact? Why do we need to run again the build? Shouldn't we download the artefact instead?

tomalec replied:
> If we generated the artifact

This is the only job in this workflow that generates an artifact.

-----

I do see we have a redundancy in general. As for example, for release PR, we run the build multiple times:
- once in general, build for bundle size on  PR trigger - https://github.com/woocommerce/google-listings-and-ads/blob/trunk/.github/workflows/build.yml#L3-L9
- once running local E2E on release PR trigger - https://github.com/woocommerce/google-listings-and-ads/blob/trunk/.github/workflows/e2e-tests.yml#L4-L46
- once here, we would also call it for release PRs.

However, to solve that, we would need to restructure our existing workflows, which is out of the scope of this PR.

We would need to make workflow interact with each other more, fetching their artifacts, controlling concurrency, etc. But I think we could benefit from taking one step back, and doing a more overall change.

:bulb: 
Suppose we redesign our workflows to distinguish reusable, composable blocks ("what") from actual flow definitions ("when _ then _"). Something like:


> reusable blocks: // What **work** should be done (technically **work**flows, jobs, & actions)
> 🏗 build package
> 👁 apply bundle watch
> 🧪 run local E2E tests
> 🆏 run QIT
> 🚀 release `gha-dev-build`
> 📖 deploy storybook
> 
> 
> work**flows**: // When what should happen - the actual flow
> - On Release PR: 🏗 , (👁, 🧪, 🆏)
> - On regular PR 🏗 , 👁
> - On push to develop 🏗 , (👁, 🧪, 🆏, 📖, 🚀)

- 
It would be easier to avoid redundancy, parallelize, and IMHO reason & maintain.
As we start to reason from the flow perspective, 
Q: "What is the process that we automate?",  
A: "release" / "regular feature development" / "develop branch consolidation" / "compatibility checks"
Q: "When do we do that?"
A: "for `release/` branch" / "for any PR" / "for pushes to `develop`" / "manual call"
Q: "What do we do?"
A: "This then that"

Instead of what we have currently:
```
run build, bundle watch, and conditional `gha-dev-build` publish on pushes and all pull requests
run storybook on pushes to develop & trunk
run build & E2E on pushes and PRs to `release/*` & manual calls
run build & QIT on manual calls, pushes to develop, and release PRs
```
When we reason only from the perspective of job, that's done. 
Q: "What do we do?"
A: "Build the package to watch its size & optionally publish it" / "Build the package to run E2E tests" / ..
Q: "When do we do that?"
A: "On releases, but also on regular pushes, but then we do only part of it. We also do the thing when we do that other thing..." :confused: 
Q: "What is the process that we automate?"
A: ":thinking: ... work?"

But that's a discussion and project on its own :)

puntope replied:
Hey @tomalec sorry maybe I expressed wrong my question.

I see this PR generates an artifact in the build workflow.

Why is this for? 

My assumption was that it was needed for the new Qit workflow... but I see we are running the full build again instead of downloading the artifact.

tomalec replied:
`woocommerce/grow/.github/workflows/run-qit-extension.yml@add/qit-workflows` 
does not run the build, it downloads the artifact :confused: 

https://github.com/woocommerce/grow/pull/81/files#diff-f653f2a856040f6c20d52799af46664b5dc76e12be91f8443c4f04f22bdf052dR78-R82

The workflow here `.github/workflows/run-qit.yml`:

0. On manual trigger
1. Runs the build workflow (`.github/workflows/build.yml`)
2. Runs the QIT which uses the artifact (`woocommerce/grow/.github/workflows/run-qit-extension.yml`)

Then we also have a case when the build is run on other triggers `.github/workflows/build.yml`:

0. On PRs and pushes to trunk/develop
1. Run the build workflow

puntope commented:
<img width="1786" alt="Screenshot 2023-09-29 at 11 58 02" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/9cd7b722-2cdc-4d21-af7c-5c98767244be">

I can see the artefact generated... however, when I download it, it generates a zip named `google-listing-and-ads.zip.zip ` so, it's double-zipped. 

<img width="573" alt="Screenshot 2023-09-29 at 11 58 14" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/b3e0bc7f-7ef1-487e-92f7-44594b2ca44f">

https://github.com/actions/upload-artifact#zipped-artifact-downloads

Not sure if this will be an issue for downloading the artefact... or we should maybe add the directory instead.

puntope commented:
I tried to run QIT workflow in the fork, and followed the steps... but I don't see it.

<img width="354" alt="Screenshot 2023-09-29 at 11 59 48" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/dd851d8f-f223-4c5a-befa-f0ff78500dd0">

What I'm doing wrong?

puntope commented:
> I tried to run QIT workflow in the fork, and followed the steps... but I don't see it.
> 
> <img alt="Screenshot 2023-09-29 at 11 59 48" width="354" src="https://user-images.githubusercontent.com/5908855/271509964-dd851d8f-f223-4c5a-befa-f0ff78500dd0.png">
> What I'm doing wrong?

Ok, I solved it by merging the branch to develop instead of trunk.

puntope commented:
I added QIT_PARTNER_SECRET and QIT_PARTNER_USER in my forked repo secrets

But I got this error

<img width="1409" alt="Screenshot 2023-09-29 at 12 15 46" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/de8990a7-1339-4c86-8c6c-3e8ea5491175">

tomalec commented:
> I added QIT_PARTNER_SECRET and QIT_PARTNER_USER in my forked repo secrets
>
> But I got this error

Could you expand the command run?

Also, did you use the CLI credentials from Pe98dy-Ir-p2 ?

tomalec committed:
Run API tests by default in run QIT workflow

 to make it consistent with manual dispatch.

 Addresses https://github.com/woocommerce/google-listings-and-ads/pull/2114#discussion_r1341004524

tomalec commented:
@puntope Thank you for the review :) would you mind another round?

puntope commented:
> Also, did you use the CLI credentials from Pe98dy-Ir-p2 ?

YEs, I used the "For CLI and GitHub Workflows" secret in the mentioned P2

puntope commented:
> Could you expand the command run?

<img width="1063" alt="Screenshot 2023-10-02 at 13 06 05" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/7608cfc8-a097-4e82-9efb-74d02ca4ed00">

<img width="1739" alt="Screenshot 2023-10-02 at 13 05 11" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/87fa3b5a-0052-4c25-840c-e0daf9a82e18">

This is with everything expanded, for some reason is not getting the credentials. This is how I set the action secrets

 
<img width="1310" alt="Screenshot 2023-10-02 at 13 07 15" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/e41d6349-c211-4828-b01d-a437f4f84ca8">

I take as granted they should be like this as the credentials is set as inherit, so I checked the GLA credentials section to see the names.

eason9487 committed:
Use the `run-qit-extension` action instead of workflow to run QIT tests and also add the artifact download processing.

tomalec committed:
Use latest test branch of QIT actions

tomalec commented:
I updated the PR, to use the latest version of grow action (instead of a workflow), so it is ready for another round.

puntope commented:
Hi @tomalec I tested again but got this error

<img width="1187" alt="Screenshot 2024-01-31 at 11 20 34" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/429284c7-ab45-4036-80b6-2385ff79246e">

puntope commented on .github/workflows/run-qit.yml:
```diff
@@ -0,0 +1,85 @@
+        uses: actions/download-artifact@v3
+        with:
+          name: google-listings-and-ads.zip
+      - name: Delegate QIT Tests
```
💅  Even if it's true that we delegate the QIT Test to the Grow Repo... I would just call this `Run QIT Tests` as it's the final goal of what we are doing. The same way we do for other workflows. We don't call them Delegate Get Woocommerce Matrix or Delegate Run PHP tests.

puntope reviewed (approved):
Hi @tomalec. This seems like working ✅  so I'm approving it in advance taking these points into consideration. 

- I left one 💅 suggestion  about naming
- ⚠️  We still need to remove the lines mentioned in the PR description before merging
- 💅  We might wait for https://github.com/woocommerce/grow/pull/81 to be merged and update the branch reference

puntope commented:
p.s Notice the phpstan checks are failing tho...

<img width="1711" alt="Screenshot 2024-01-31 at 11 53 01" src="https://github.com/woocommerce/google-listings-and-ads/assets/5908855/66ec6a6f-caf6-4efb-a564-14a5755bb53b">

tomalec committed:
Rename qit workflow step name

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/2114#discussion_r1473135632

tomalec commented:
Thank you for the review :)

>p.s Notice the phpstan checks are failing tho...

This is kinda expected, as GLA fails to pass those.

1. I pushed your suggestion.
2. I'll update the references, now that it is reviewed.
3. Yep, let's wait.

tomalec committed:
Remove testing lines from QIT workflow,

prepare for the released grow action
